### PR TITLE
[Compat Api] Create System Icon Abstraction

### PR DESCRIPTION
**Pull request by camilaibs**
## Hey, I just made a Pull Request!

Create a compact system icon component to address the following comment:
https://github.com/backstage/backstage/pull/22752#discussion_r1482775935

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: packages/core-compat-api/src/components/SystemIcon.tsx [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+/**
+ * @public
+ * Props for the System Icon component.
+ */
+export type SystemIconProps = {
+  // The id of the system icon to render.
+  id: string;
+  // An optional fallback element to render when the system icon is not found.
+  fallback?: JSX.Element;
```
Comment (Rugvip) [inline packages/core-compat-api/src/components/SystemIcon.tsx:29]: Wanna hear your thoughts on this element approach vs `IconComponent` 😁

Could be the case that we want the `SystemIcon` to accept a `fontSize` prop to set the size, I think that's fairly common. Might be that the best approach for that is to always set an outer size and let the icon inherit it though.

**Review by Rugvip — COMMENTED** [review]
Nice! 👍 

Few early comments

File: packages/core-compat-api/src/components/SystemIcon.tsx [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+ * @public
+ * SystemIcon is a component that renders a system icon by its id.
+ * @example
+ * Rendering the "kind:api" icon:
+ * ```tsx
+ * <SystemIcon id="kind:api" />
+ * ```
+ */
+function CompatSystemIcon(props: SystemIconProps) {
+  return compatWrapper(<SystemIcon {...props} />);
```
[the remark this answers is not in the document]
Comment (camilaibs) [reply packages/core-compat-api/src/components/SystemIcon.tsx:49]: Yeah, I a added it so the useApp would work independently of extra wrapping... Could this be wrapped in a try-catch and used only if the use app throws an error 🤔 ?

File: packages/core-compat-api/src/components/SystemIcon.tsx (hunk 1 above)
Comment (camilaibs) [reply packages/core-compat-api/src/components/SystemIcon.tsx:29]: I forgot about this `IconComponent` 😄 . I guess we can accept all `IconComponentProps` and forward them to the returned system component or fallback... I'll try it. Thanks!

Comment (camilaibs) [reply packages/core-compat-api/src/components/SystemIcon.tsx]: https://github.com/backstage/backstage/pull/22892/commits/d5f859bd8749c71f0d6fa03220edd4007ba0ac62

File: packages/core-compat-api/src/components/SystemIcon.tsx [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  return <Component {...rest} />;
+}
+
+/**
+ * @public
+ * SystemIcon is a component that renders a system icon by its id.
+ * @example
+ * Rendering the "kind:api" icon:
+ * ```tsx
+ * <SystemIcon id="kind:api" />
```
Comment (aramissennyeydd) [inline packages/core-compat-api/src/components/SystemIcon.tsx:46]: One thing I've wanted to do for a while with the existing icons was to allow differentiation by `spec.type`, kind=Resource and kind=Component components aren't super helpful in the catalog graph. I ended up sticking more information into the entityRef/name for visibility when a better icon granularity might have sufficed. Seems like there's a similar ask for search, #12876. Is that something we could do with this wrapper? Allowing either an entity/entityRef as input or a more specific id?

Comment (aramissennyeydd) [reply packages/core-compat-api/src/components/SystemIcon.tsx:46]: Should the presentation API use this API for icons or are these separate concepts?

Comment (camilaibs) [reply packages/core-compat-api/src/components/SystemIcon.tsx:46]: My understanding is that for me it is a temporary compatibility component. We have a new Icons API and also a new Components API, so my initial understanding is that at some point we will deprecate this component and switch to using one of these two APIs. But let me add @Rugvip to the conversation as well to confirm 🙂

Comment (Rugvip) [reply packages/core-compat-api/src/components/SystemIcon.tsx:46]: Ah no the idea is actually for this component to stick around. It's just a lot simpler to use this than having to juggle the icons api in many places.

I wonder if what we actually need here is the ability to specify some fallbacks or to have that be some implicit behavior? For example one might specify `entity:component:service`, `entity:component`, and `entity`, and it uses the first one that is available? Or for example colons have special treatment so that specifying `entity:component:service` will look for `entity:component` and `entity` too?

Comment (camilaibs) [reply packages/core-compat-api/src/components/SystemIcon.tsx:55]: Just updated the logic, could you review again?

File: packages/core-compat-api/src/components/SystemIcon.test.tsx [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+import { renderInTestApp } from '@backstage/test-utils';
+import { SystemIcon } from './SystemIcon';
+
+describe('SystemIcon', () => {
+  it('should render the correct system icon', async () => {
+    const { container } = await renderInTestApp(<SystemIcon keys="kind:api" />);
+    expect(container.querySelector('svg')).toBeDefined();
+  });
+
+  it('should render the first found ico when multiple keys are provided', async () => {
```
Comment (aramissennyeydd) [inline packages/core-compat-api/src/components/SystemIcon.test.tsx:28]:
```suggestion
  it('should render the first found icon when multiple keys are provided', async () => {
```

**Review by aramissennyeydd — APPROVED** [review]
Nice 🎉 ! At first glance the array of keys feels a little odd, but I think it's exactly what I needed 😁

File: packages/core-compat-api/src/components/SystemIcon.tsx (hunk 3 above)
Comment (aramissennyeydd) [reply packages/core-compat-api/src/components/SystemIcon.tsx]: Just to clarify, this component won't / shouldn't be used for the presentation API implementation? This component will be deprecated soon and I'd have to do something similar myself?

Comment (aramissennyeydd) [reply packages/core-compat-api/src/components/SystemIcon.tsx]: > As a matter of fact, the kind icons map can be used as input to the default renderer, but that's a separate issue and just a historical artifact. Making a system icon renderer that then consumes through the presentation api ... sounds like a weird circular situation :)

My proposal is exactly that, moving away from the kind icons map towards using `SystemIcon` for everything. Maybe I'm just confused by the use of `kind:api` as both a system icon and an entity icon. Seems like there's overlap that exists already that we could leverage?

File: packages/core-compat-api/src/components/SystemIcon.tsx [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+import { useApp, IconComponent } from '@backstage/core-plugin-api';
+import { compatWrapper } from '../compatWrapper';
+
+/**
+ * @public
+ * Props for the SystemIcon component.
+ */
+export type SystemIconProps = ComponentProps<IconComponent> & {
+  // The id of the system icon to render, if provided as an array, the first icon found will be rendered.
+  keys: string | string[];
```
Comment (Rugvip) [inline packages/core-compat-api/src/components/SystemIcon.tsx:27]: Ah in the end I don't think we really wanna go with this one either. I think there's a pretty big risk that being able to list multiple keys ends up being really confusing to adopters that are configuring icons. It's kinda bad enough that adding icons for certain keys can replace icons in other plugins, but if you also need to juggle the case where there may be multiple different keys that affect certain plugins and try to find the correct key to use for a certain update, it all gets really messy.

tl;dr: thinking just `key: string` for now

File: packages/core-compat-api/src/components/SystemIcon.tsx [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+ * ```tsx
+ * <SystemIcon keys="kind:api" Fallback={ApiFallbackIcon} />
+ * ```
+ * @example
+ * Customizing the icon font size:
+ * ```tsx
+ * <SystemIcon keys="kind:api" fontSize="medium" />
+ * ```
+ */
+function CompatSystemIcon(props: SystemIconProps) {
```
Comment (Rugvip) [inline packages/core-compat-api/src/components/SystemIcon.tsx:67]: I actually had in mind that this isn't just a compat thing, but new API added to `@backstage/frontend-plugin-api` to make it easier to render icons. Having though about it some more I'll backtrack that a little bit to not make it part of the core API, but I'm thinking that adding this to `@backstage/core-components` instead would make sense? Basically as a replacement for all of these: https://github.com/backstage/backstage/blob/e643a59dca04a24cebe661c62c62fb4c1f75bd17/packages/core-components/src/icons/icons.tsx#L36-L78

I'm thinking the best name would be `AppIcon`, but let me know what you think.

Comment (camilaibs) [reply packages/core-compat-api/src/components/SystemIcon.tsx]: Good idea, pull request updated 🙌🏻

File: packages/core-compat-api/src/components/SystemIcon.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (camilaibs) [reply packages/core-compat-api/src/components/SystemIcon.tsx]: Not used anymore.

**Review by Rugvip — APPROVED** [review]
:shipit: 😎
